### PR TITLE
Use a better system for diagnostics with debouncing

### PR DESCRIPTION
**Pull request by jmbockhorst**
Use a better system for handling diagnostics. Only send a diagnostics message if the files diagnostics have changed and also debounce requests. Early testing seems like this improves performance by a lot in large workspaces.

### CODE REVIEW

File: src/providers/diagnostics/diagnosticsProvider.ts [hunk 1]
```diff
@@ -28,18 +35,55 @@ export interface IElmIssue {
...
 }
 
+const enum DiagnosticKind {
+  ElmMake,
+  ElmAnalyse,
+  ElmTest,
+  TypeInference,
+}
+
+function rangeEquals(a: Range, b: Range): boolean {
```
Comment (Razzeee) [inline src/providers/diagnostics/diagnosticsProvider.ts:45]: I guess we should move this to a general place too, calling this should be helpful in other files.

File: src/providers/diagnostics/diagnosticsProvider.ts [hunk 2]
```diff
@@ -140,75 +173,63 @@ export class DiagnosticsProvider {
...
-      }
+    if (fileDiagnostics) {
+      didUpdate = fileDiagnostics.update(kind, diagnostics);
+    } else if (diagnostics.length > 0) {
+      fileDiagnostics = new FileDiagnostics(uri);
+      fileDiagnostics.update(kind, diagnostics);
+      this.currentDiagnostics.set(uri, fileDiagnostics);
+      didUpdate = true;
     }
+    const sendDiagnostics = (uri: string): void => {
```
Comment (Razzeee) [inline src/providers/diagnostics/diagnosticsProvider.ts:210]: Do we really need to construct these when `didUpdate` is not true?

File: src/providers/diagnostics/diagnosticsProvider.ts [hunk 3]
```diff
@@ -254,7 +284,37 @@ export class DiagnosticsProvider {
     ) {
       await this.elmAnalyseDiagnostics.updateFile(uri, text);
     }
+  }
+}
+
+class FileDiagnostics {
```
Comment (Razzeee) [inline src/providers/diagnostics/diagnosticsProvider.ts:290]: Move this to it's own file please :)

File: src/providers/diagnostics/diagnosticsProvider.ts [hunk 4]
```diff
@@ -28,18 +35,55 @@ export interface IElmIssue {
...
+function rangeEquals(a: Range, b: Range): boolean {
+  return (
+    a.start.character === b.start.character &&
+    a.start.line === b.start.line &&
+    a.end.character === b.end.character &&
+    a.end.line === b.end.line
+  );
+}
+
+function diagnosticsEquals(a: Diagnostic, b: Diagnostic): boolean {
```
Comment (Razzeee) [inline src/providers/diagnostics/diagnosticsProvider.ts:54]: Should be moved out together with FileDiagnostics

File: src/providers/diagnostics/diagnosticsProvider.ts [hunk 5]
```diff
@@ -254,7 +284,37 @@ export class DiagnosticsProvider {
...
+    return [
+      ...this.getForKind(DiagnosticKind.ElmMake),
+      ...this.getForKind(DiagnosticKind.ElmAnalyse),
+      ...this.getForKind(DiagnosticKind.ElmTest),
+      ...this.getForKind(DiagnosticKind.TypeInference),
+    ];
+  }
+
+  public update(kind: DiagnosticKind, diagnostics: Diagnostic[]): boolean {
+    const existing = this.diagnostics.get(kind);
```
Comment (Razzeee) [inline src/providers/diagnostics/diagnosticsProvider.ts:308]: Using `getForKind` might make sense, then you can drop the existing check in the next line

**Review by Razzeee — APPROVED** [review]
Code looks good to me, will have a runtime test and then merge.

**Comment by jmbockhorst** [comment]
Cool. I also tested during work yesterday afternoon and didn't see anything unexpected.

**Pull request merged**